jormungandr/src/blockchain/process.rs

```diff
@@ -401,29 +298,91 @@ fn try_request_fragment_removal(
 /// this function will re-process the tip against the different branches
 /// this is because a branch may have become more interesting with time
 /// moving forward and branches may have been dismissed
-pub fn reprocess_tip(
-    logger: Logger,
-    blockchain: Blockchain,
-    tip: Tip,
-) -> impl Future<Item = (), Error = Error> {
-    let branches_future = blockchain.branches().branches();
-
-    branches_future
-        .join(tip.get_ref())
-        .map(|(all, tip)| {
-            all.into_iter()
-                .filter(|r|
-                    // remove our own tip so we don't apply it against itself
-                    !Arc::ptr_eq(&r, &tip))
-                .collect::<Vec<_>>()
-        })
-        .and_then(move |others| {
-            stream::iter_ok(others)
-                .for_each(move |other| {
-                    process_new_ref(logger.clone(), blockchain.clone(), tip.clone(), other)
-                })
-                .into_future()
-        })
+async fn reprocess_tip(logger: Logger, mut blockchain: Blockchain, tip: Tip) -> Result<(), Error> {
+    let branches: Vec<Arc<Ref>> = blockchain.branches().branches().compat().await.unwrap();
+
+    let tip_as_ref = tip.get_ref_std().await;
+
+    let others = branches
+        .iter()
+        .filter(|r| !Arc::ptr_eq(&r, &tip_as_ref))
+        .collect::<Vec<_>>();
+
+    for other in others {
+        process_new_ref_std(&logger, &mut blockchain, tip.clone(), Arc::clone(other)).await?
+    }
+
+    Ok(())
+}
+
+/// process a new candidate block on top of the blockchain, this function may:
+///
+/// * update the current tip if the candidate's parent is the current tip;
+/// * update a branch if the candidate parent is that branch's tip;
+/// * create a new branch if none of the above;
+///
+/// If the current tip is not the one being updated we will then trigger
+/// chain selection after updating that other branch as it may be possible that
+/// this branch just became more interesting for the current consensus algorithm.
+pub async fn process_new_ref_std(
+    logger: &Logger,
+    blockchain: &mut Blockchain,
+    mut tip: Tip,
+    candidate: Arc<Ref>,
+) -> Result<(), Error> {
+    let candidate_hash = candidate.hash();
+    let storage = blockchain.storage().clone();
+
+    let tip_ref = tip.get_ref_std().await;
+
+    let tip_updated = if tip_ref.hash() == candidate.block_parent_hash() {
+        info!(
+            logger,
+            "update current branch tip: {} -> {}",
+            tip_ref.header().description(),
+            candidate.header().description(),
```

Comment:
you are repeating yourself with the logs here. 

create a log with the new context value instead. I.e:

```rust
let logger = logger.new(o!(
  "current_tip" => tip_ref.header().description(),
  "candidate" => tip.header().description(),
));
``` 

That will make the code shorter too and more easily readable.

Reply:
just converting the old code, not doing any substantial improvement or anything different right now